x-pack/plugins/proxy/server/proxy.ts

```diff
@@ -0,0 +1,291 @@
+}
+
+export interface RoutingNode {
+  type: string; // what are all the types this can be?
+  node: string;
+  state: RouteState;
+}
+
+interface LivenessNode {
+  lastUpdate: number;
+  lastChangedTime: number;
+}
+
+interface ClusterDoc {
+  nodes: {
+    [key: string]: LivenessNode;
+  };
+  routing_table: {
+    [key: string]: RoutingNode;
+  };
+}
+
+export type RoutingTable = Map<string, RoutingNode>;
+type NodeList = Map<string, LivenessNode>;
+
+// convert a map<string, V> to a plain javascript object of {[key:string]: val}
+function map2obj<V>(map: Map<string, V>): { [key: string]: V } {
+  return [...map].reduce((a, [k, v]) => Object.assign(a, { [k]: v }), {});
+}
+
+export class ProxyService {
+  private updateInterval: number = 1000 * 15;
+  private timeoutThreshold: number = 1000 * 30;
+  private port: number = 5602;
+  private timer: null | NodeJS.Timeout = null;
+  private configSubscription?: Subscription;
+  private routingTable: RoutingTable = new Map();
+  private elasticsearch?: Observable<ClusterClient>;
+  private proxyServer?: HttpServerSetup;
+  private nodeName: string;
+
+  private readonly proxyIndex = '.kibana';
+  private readonly proxyDoc = 'proxy-resource-list';
+  private readonly httpsAgent: HTTPSAgent;
+  private readonly httpAgent: HTTPAgent;
+  private readonly allowUnauthAgent: HTTPAgent;
+  private readonly wreck: typeof Wreck;
+  private readonly log: Logger;
+  private readonly config$: Observable<ProxyConfig>;
+
+  constructor(initializerContext: PluginInitializerContext) {
+    this.nodeName = v4();
+    this.config$ = initializerContext.config.create<ProxyConfig>();
+    this.log = initializerContext.logger.get('proxy');
+    this.httpsAgent = new HTTPSAgent({ keepAlive: true });
+    this.httpAgent = new HTTPAgent({ keepAlive: true });
+    this.allowUnauthAgent = new HTTPAgent({ keepAlive: true });
+    this.wreck = Wreck.defaults({
+      agent: {
+        https: this.httpsAgent,
+        http: this.httpAgent,
+        httpsAllowUnauthorized: this.allowUnauthAgent,
+      },
+    });
+  }
+
+  public async setup(core: CoreSetup) {
+    // initialize cluster information
+    // we need to get this kibana instance's ip address
+    // it's hostname and determine which port to listen on
+    // should liveness timing be part of configuration?o
+    // mikhael -- talk about changing http service to allow multiple services
+    this.configSubscription = this.config$.subscribe(config => {
+      this.setConfig(config);
+    });
+
+    const config = await this.config$.pipe(first()).toPromise();
+    this.setConfig(config);
+    this.elasticsearch = core.elasticsearch.dataClient$;
+    this.proxyServer = await core.http.createNewServer({ port: config.port });
+  }
+
+  private setConfig(config: ProxyConfig) {
+    this.updateInterval = config.updateInterval || this.updateInterval;
+    this.timeoutThreshold = config.timeoutThreshold || this.timeoutThreshold;
+    this.port = config.port || this.port;
```

Comment:
config marks `port` as required, you will never fallback to this.port, probably it should be removed from the class